docs/proposals/13-automated-seed-management.md

```diff
@@ -0,0 +1,369 @@
+
+Later, the scheduling algorithm could be further enhanced by replacing the step in which the region strategy is applied by a scoring step similar to the one in [Kubernetes Scheduler](https://kubernetes.io/docs/concepts/scheduling-eviction/kube-scheduler/). In this scoring step, the scheduler would rank the remaining seeds to choose the most suitable shoot placement. It would assign a score to each seed that survived filtering based on a list of scoring rules. These rules might include for example `MinimalDistance` and `SeedResourcesLeastAllocated`, among others. Each rule would produce its own score for the seed, and the overall seed score would be calculated as a weighted sum of all such scores. Finally, the scheduler would assign the shoot to the seed with the highest ranking. 
+
+## Creating, Deleting, and Updating Seeds Declaratively
+
+When all or most of the existing seeds are near capacity, new seeds should be created in order to accommodate more shoots. Conversely, sometimes there could be too many seeds for the number of shoots, and so some of the seeds could be deleted to save resources. Currently, the process of creating a new seed involves a number of manual steps, such as creating a new shoot that meets certain criteria, and then registering it as a seed in Gardener. This could be automated to some extent by [annotating a shoot with the `use-as-seed` annotation](../usage/shooted_seed.md), in order to create a "shooted seed". However, adding more than one similar seeds still requires manually creating all needed shoots, annotating them appropriately, and making sure that they are successfully reconciled and registered.
+
+To create, delete, and update seeds effectively in a declarative way and allow auto-scaling, a "creatable seed" resource along with a "set" (and in the future, perhaps also a "deployment") of such creatable seeds should be introduced, similar to Kubernetes `Pod`, `ReplicaSet`, and `Deployment` (or to MCM `Machine`, `MachineSet`, and `MachineDeployment`) resources. With such resources (and their respective controllers), creating a new seed based on a template would become as simple as increasing the `replicas` field in the "set" resource. 
+
+In [Issue #2181](https://github.com/gardener/gardener/issues/2181) it is already proposed that the `use-as-seed` annotation is replaced by a dedicated `ShootedSeed` resource. The solution proposed here further elaborates on this idea.
+
+### ShootedSeeds
+
+#### ShootedSeed Resource
+
+The `ShootedSeed` resource is a dedicated custom resource that represents a "shooted seed" and properly replaces the `use-as-seed` annotation. This resource contains:
```

Comment:
From my point of view, the reason is that you want to scale shooted seeds later on. The idea is to create `ShootedSeedSet` resources that contain the shoot spec template. When changing the `replicas` then a `ShootedSeed` will be created which itself results in the creation of the `Shoot`.
What you propose @timebertt would mean that when you change the `replicas` of `ShootedSeedSet` then it creates a `Shoot` first (based on the spec template), and then a `ShootedSeed` that refers to this `Shoot`. Isn't this more confusing and contrary to what people are used from other Kubernetes resources (`ReplicaSet`, `MachineSet`)?

Reply:
I would be very happy to get rid of the complexity, however, how do you manage such ShootedSeeds as part of a ShootedSeedSet? They obviously can't refer to the same shoot, nether could they refer to different existing shoots. The issue with ShootedSeeds defined in this way is that you can't easily create more than one replica, and therefore it can't be used in a "set" resource.

Come to think of it, perhaps this could be solved by adding a `shootTemplate` in addition to the `template` to the `ShootedSeedSet` spec. To create a replica, the ShootedSeedSet controller would need to first create a Shoot from `shootTemplate`, and then a ShootedSeed from `template` using the previously created Shoot. But then the logic I initially intended to put in the ShootedSeed controller would migrate to the ShootedSeedSet controller, leading to similar complexity and challenges. I would rather prefer the ShootedSeed as currently proposed.

BTW, some of the benefits you list are already achieved by introducing the `ShootedSeedRegistration`. If we didn't have the "set" in mind, we could simply stop here and all would be perfect, but unfortunately we can't do this.

What do you think?

Reply:
@rfranzke Thanks for your comments, for whatever reason I did not see them when I replied yesterday. I am thinking hard about this since yesterday, and the more I think the more I like @timebertt idea above. 
* In fact, there is an example in Kubernetes about a similar arrangement - the [stateful set](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) manages a number of pods based on a template and an equal number of matching PVCs based on a PVC template.
* The main advantage that I see compared to a ShootedSeed as originally proposed is that a Set doesn't try to maintain the objects it has created, once they are created. This means that the Shoots created by the Set can be changed by other GCM controllers, scheduled by the scheduler, etc., and the Set would not try to overwrite such changes by constant reconciliation. This is not the case with the ShootedSeed containing a Shoot template - there, all changes must be maintained in the original ShootedSeed, resulting in changes to existing GCM controllers and the scheduler, as pointed out already in [Changes to Existing Controllers](https://github.com/stoyanr/gardener/blob/add-seed-managment-proposal/docs/proposals/13-automated-seed-management.md#changes-to-existing-controllers).
In short, with this idea we would achieve the benefits that @timebertt lists above to a greater extent, which makes it worthwhile in my opinion.

Reply:
Yeah, this is a bit tricky, I see that both approaches have their pros and cons.
I don't have a strong opinion on which approach to choose yet. So we can think through all implications of both approaches in the in-person discussion and then it might be easier to decide.

Reply:
In an in-person discussion in which @rfranzke also participated, we agreed that:

* There should be only one resource named `ManagedSeed` with the structure of what was initially proposed as `SeedRegistration`.
* The `ManagedSeedSet` will contain also a shoot template.

The document has been already updated to reflect the above.